### PR TITLE
Test parts of API with libFuzzer.

### PR DESCRIPTION
**Pull request by PatKamin**
Add a fuzz test generating API calls. Related to #566 . 
Replaces #731 (changed the target branch to adapters).

### CODE REVIEW

File: test/fuzz/utils.hpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    UR_PROGRAM_CREATE_WITH_IL,
+    UR_MAX_FUZZER_API_CALL,
+};
+
+typedef struct FuzzerInput {
+    uint8_t *data_ptr;
+    size_t data_size;
+} FuzzerInput;
+
+typedef struct TestState {
```
Comment (pbalcer) [inline test/fuzz/utils.hpp:41]: do you need to use typedef like in C?

File: test/fuzz/utils.hpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        return -1;
+    }
+    *out_data = *input->data_ptr;
+    input->data_ptr += out_data_size;
+    input->data_size -= out_data_size;
+
+    return 0;
+}
+
+int init_random_data(TestState *state) {
```
Comment (pbalcer) [inline test/fuzz/utils.hpp:80]: it's not really random, right? it's just reading N values from the input to create initial state.

File: test/fuzz/utils.hpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    if (state->input.data_size < 5) {
```
Comment (pbalcer) [inline test/fuzz/utils.hpp:81]: hm, wouldn't it be better to check result of `get_next_input_data`? otherwise, if the size of these fields changes, this will not be correct.

File: test/fuzz/utils.hpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        return -1;
+    }
+    get_next_input_data(&state->input, &state->platform_num);
+    get_next_input_data(&state->input, &state->device_num);
+    get_next_input_data(&state->input, &state->context_num);
+    get_next_input_data(&state->input, &state->device_type_fuzz);
+    if (state->device_type_fuzz < 1 || state->device_type_fuzz > 7) {
```
Comment (pbalcer) [inline test/fuzz/utils.hpp:88]: `UR_DEVICE_TYPE_MAX` or something?

File: test/fuzz/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    add_ur_executable(${TEST_TARGET_NAME}
+        ${ARGN})
+    target_link_libraries(${TEST_TARGET_NAME}
+        PRIVATE
+        ${PROJECT_NAME}::loader
+        ${PROJECT_NAME}::headers
+        ${PROJECT_NAME}::common
+        -fsanitize=fuzzer -fprofile-instr-generate -fcoverage-mapping)
+    add_test(NAME ${TEST_TARGET_NAME}
+        COMMAND ${TEST_TARGET_NAME} -max_total_time=600 -seed=1 -shrink=1 -verbosity=1
```
Comment (pbalcer) [inline test/fuzz/CMakeLists.txt:17]: We should have a set of predefined input data files we know pass for CI, and a separate setup where we run the fuzzer for longer periods of time. Even with a fixed seed, running this on every build might make the CI flaky.

File: test/fuzz/urFuzz.cpp [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+        return -1;
+    }
+
+    ur_usm_pool_handle_t pool;
+    ur_usm_pool_desc_t pool_desc{UR_STRUCTURE_TYPE_USM_POOL_DESC, nullptr,
+                                 UR_USM_POOL_FLAG_ZERO_INITIALIZE_BLOCK};
+    ur_result_t res =
+        urUSMPoolCreate(state.contexts[state.context_num], &pool_desc, &pool);
+    if (res == UR_RESULT_SUCCESS) {
+        pool_allocs[pool] = {};
```
Comment (pbalcer) [inline test/fuzz/urFuzz.cpp:101]: pool_allocs.insert(pool) ?

File: test/fuzz/utils.hpp (hunk 1 above)
Comment (pbalcer) [inline test/fuzz/utils.hpp:41]: wouldn't it be simpler to make the util functions part of this struct?

File: test/fuzz/utils.hpp (hunk 2 above)
Comment (PatKamin) [reply test/fuzz/utils.hpp:80]: It's random in a sense that the inupt is generated by a fuzzer. Of course, the fuzzer will probably stick to some values more than the others. Should I replace it with generating random values with the help of a random header?

File: test/fuzz/utils.hpp (hunk 3 above)
Comment (PatKamin) [reply test/fuzz/utils.hpp:81]: I replaced multiple checks after each and every `get_next_input_data` with this look-ahead check for all values at once. I will change it though, as one have to remember to change this value every time a "random" var is added/removed.

File: test/fuzz/CMakeLists.txt (hunk 5 above)
Comment (PatKamin) [reply test/fuzz/CMakeLists.txt:17]: OK, and in the predefined input the fuzzer should provide just functions parameters and not generate API calls on the fly, right?

File: test/fuzz/utils.hpp (hunk 2 above)
Comment (pbalcer) [reply test/fuzz/utils.hpp:80]: No, my point is that the name isn't really accurate. I'd just call it init_state or something.

File: test/fuzz/CMakeLists.txt (hunk 5 above)
Comment (pbalcer) [reply test/fuzz/CMakeLists.txt:17]: yes, it should be a file that contains all the bytes you use in your program to create some sequence of operations.

File: test/fuzz/utils.hpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+    uint8_t platform_num;
+    uint8_t device_num;
+    uint8_t context_num;
+    uint8_t device_type_fuzz;
+} TestState;
+
+//////////////////////////////////////////////////////////////////////////////
+template <typename T> int get_next_input_data(FuzzerInput *input, T *out_data) {
+    size_t out_data_size = sizeof(out_data);
```
Comment (igchor) [inline test/fuzz/utils.hpp:69]: shoudln't it be `sizeof(*out_data)`?

File: test/fuzz/utils.hpp [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    uint32_t num_devices;
+
+    uint8_t platform_num;
+    uint8_t device_num;
+    uint8_t context_num;
+    uint8_t device_type_fuzz;
+} TestState;
+
+//////////////////////////////////////////////////////////////////////////////
+template <typename T> int get_next_input_data(FuzzerInput *input, T *out_data) {
```
Comment (igchor) [inline test/fuzz/utils.hpp:68]: I'm not sure if you need to write this by yourself. There are already some untils in llvm to split the fuzzing input into multiple 'sub-inputs': https://github.com/google/fuzzing/blob/master/docs/split-inputs.md For me it looks you can just use FuzzedDataProvider  (https://raw.githubusercontent.com/llvm/llvm-project/main/compiler-rt/include/fuzzer/FuzzedDataProvider.h) by calling ConsumeIntegral()/ConsumeEnum, etc.

Comment (igchor) [reply test/fuzz/utils.hpp:68]: And even if you need to write something by yourself I would recommend just wrapping FuzzedDataProvider.ConsumeBytes()

File: test/fuzz/utils.hpp [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+    return 0;
+}
+
+int get_next_api_call(FuzzerInput *input) {
+    uint8_t next_api_call;
+    if (get_next_input_data(input, &next_api_call) != 0) {
+        return -1;
+    }
+    return next_api_call % UR_MAX_FUZZER_API_CALL;
```
Comment (igchor) [inline test/fuzz/utils.hpp:101]: See my comment above, you can probably just use FuzzedDataProvider::ConsumeEnum<ApiCall> or something like this. (where ApiCall is an enum describing the call)

File: test/fuzz/utils.hpp [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+        state->devices[0] == nullptr) {
+        return false;
+    }
+
+    return true;
+}
+
+bool check_context_exists(const TestState *state) {
+    if (state->contexts.empty() ||
+        state->context_num >= state->contexts.size() ||
```
Comment (igchor) [inline test/fuzz/utils.hpp:115]: instead of checking `state->context_num >= state->contexts.size()` I think it would be better to just generate such value that is < size(). Unless, we want to fuzz cases where this number is bigger but perhaps that should be just a special-test case.

File: test/fuzz/utils.hpp [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        state->contexts[0] == nullptr) {
+        return false;
+    }
+
+    return true;
+}
+
+bool check_pool_exists(
```
Comment (igchor) [inline test/fuzz/utils.hpp:123]: Same comment here, can't we just generate a pool_num value that is < map.size()?

File: test/fuzz/utils.hpp [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+
+int get_next_api_call(FuzzerInput *input) {
+    uint8_t next_api_call;
+    if (get_next_input_data(input, &next_api_call) != 0) {
+        return -1;
+    }
+    return next_api_call % UR_MAX_FUZZER_API_CALL;
+}
+
+bool check_device_exists(const TestState *state) {
```
Comment (igchor) [inline test/fuzz/utils.hpp:104]: Same comment here (as below).

File: test/fuzz/utils.hpp [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+int load_kernel_source(std::vector<char> &binary_out) {
+    std::string source_path = KERNEL_IL_PATH;
+
+    std::ifstream source_file;
+    source_file.open(source_path,
+                     std::ios::binary | std::ios::in | std::ios::ate);
+    if (!source_file.is_open()) {
+        std::cerr << "Failed to open a kernel source file: " << source_path
+                  << std::endl;
+        return -1;
```
Comment (igchor) [inline test/fuzz/utils.hpp:170]: why not just throw an exception and return a vector? there are probably already some unhandled exceptions in this code (perhaps from open())

File: test/fuzz/utils.hpp [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+    }
+    source_file.close();
+
+    binary_out = std::vector<char>(std::move(device_binary));
+
+    return 0;
+}
+
+void cleanup(TestState *state) {
+    urLoaderConfigRelease(state->config);
```
Comment (igchor) [inline test/fuzz/utils.hpp:192]: nit: Can we just wrap all the things that need to be releaed/freed/closed in unique_ptrs instead of implementing this function?

Comment (PatKamin) [reply test/fuzz/utils.hpp]: Done

File: test/CMakeLists.txt [hunk 15]
```diff
@@ -24,3 +24,6 @@ add_subdirectory(unit)
 if(UR_BUILD_TOOLS)
   add_subdirectory(tools)
 endif()
+if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
```
[the remark this answers is not in the document]
Comment (PatKamin) [reply test/CMakeLists.txt]: Thanks, done

File: test/fuzz/utils.hpp [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+    uint8_t platform_num;
+    uint8_t device_num;
+    uint8_t context_num;
+
+    TestState(std::unique_ptr<FuzzedDataProvider> data_provider)
+        : data_provider(std::move(data_provider)) {}
+
+    template <typename IntType> int get_next_input_data(IntType *data) {
+        if (data_provider->remaining_bytes() < sizeof(IntType)) {
```
Comment (igchor) [inline test/fuzz/utils.hpp:101]: I'm wondering if checking for remaining_bytes here and returning -1 makes sense. Why not just always call `consume*`, even if there are no bytes left? From the documentation, it will just return the same value every time (e.g. 0 for int). 

I don't know which approach is better for coverage but it certainly seems simpler as we don't need to deal with `-1`.

But feel free to ignore this comment. We can also experiment later with what gives better results.

File: test/fuzz/utils.hpp [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+struct Context {
+    ur_context_handle_t handle;
+    Pools host_pools;
+    Pools device_pools;
+    Allocs no_pool_host_allocs;
+    Allocs no_pool_device_allocs;
+
+    Context(ur_context_handle_t handle) : handle(handle) {}
+    ~Context() {
+        host_pools.clear();
```
Comment (igchor) [inline test/fuzz/utils.hpp:69]: nit: no need to call `.clear()`

Comment (PatKamin) [reply test/fuzz/utils.hpp:79]: It's for calling Alloc dtors before removing Context, avoiding use-after-free.

File: test/fuzz/CMakeLists.txt (hunk 5 above)
Comment (PatKamin) [reply test/fuzz/CMakeLists.txt]: Done

File: test/fuzz/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    target_compile_definitions(${TEST_TARGET_NAME} PRIVATE -DKERNEL_IL_PATH="${UR_CONFORMANCE_DEVICE_BINARIES_DIR}/bar/sycl_spir641.spv")
+    target_include_directories(${TEST_TARGET_NAME} PRIVATE ${UR_CONFORMANCE_DEVICE_BINARIES_DIR})
+
+    add_dependencies(${TEST_TARGET_NAME} generate_device_binaries)
+endfunction()
+
+add_fuzz_test(base fuzz-long
+    -max_total_time=1 -seed=1 -verbosity=1)
+
+set(CORPUS_DIR ${CMAKE_CURRENT_SOURCE_DIR}/corpus)
```
Comment (pbalcer) [inline test/fuzz/CMakeLists.txt:34]: maybe not for now, but might be useful to have something like:
```
foreach(file with some extension in corpus_dir) {
   add_test(...);
}
```

**Review by pbalcer — APPROVED** [review]
I've noticed the python test is failing for some reason. Maybe you need to rebase?

**Comment by PatKamin** [comment]
> I've noticed the python test is failing for some reason. Maybe you need to rebase?

Thanks, rebased.

File: test/fuzz/corpus/README.md [hunk 19]
```diff
@@ -0,0 +1,87 @@
+# Corpra for fuzz tests with fixed API calls scenarios
+These corpra contain UR API calls in a predefined order described below.
+All scenarios begin with single calls to urInit() and urAdapterGet().
+
```
Comment (lukaszstolarczuk) [inline test/fuzz/corpus/README.md:4]: could you please add some info how to create new/update existing corpras...? perhaps some link to a guide or something

File: test/fuzz/urFuzz.cpp [hunk 20]
```diff
@@ -0,0 +1,415 @@
+// Copyright (C) 2023 Intel Corporation
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+// See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
```
Comment (lukaszstolarczuk) [inline test/fuzz/urFuzz.cpp:5]: could you please add a short description, what is this file's purpose :wink:

File: .github/workflows/daily.yml [hunk 21]
```diff
@@ -0,0 +1,70 @@
+name: Build and test
```
Comment (lukaszstolarczuk) [inline .github/workflows/daily.yml:1]: perhaps call the file "nigthly" and here also a different name than "Build and test" (to not duplicate existing workflow name)

File: .github/workflows/cmake.yml [hunk 22]
```diff
@@ -63,6 +65,10 @@ jobs:
         wget -O ${{github.workspace}}/dpcpp_compiler.tar.gz https://github.com/intel/llvm/releases/download/sycl-nightly%2F20230626/dpcpp-compiler.tar.gz
         tar -xvf ${{github.workspace}}/dpcpp_compiler.tar.gz
 
+    - name: Setup DPC++
```
Comment (lukaszstolarczuk) [inline .github/workflows/cmake.yml:68]: why do we need this, actually? if it's only needed for fuzzing, please add an `if` here (and perhaps add a comment on the purpose of this)

Comment (PatKamin) [reply .github/workflows/cmake.yml:66]: It's needed for building kernel binaries with sycl. Removed from the path not running fuzztests.

File: .github/workflows/daily.yml (hunk 21 above)
Comment (PatKamin) [reply .github/workflows/daily.yml]: Done

File: test/fuzz/urFuzz.cpp (hunk 20 above)
Comment (PatKamin) [reply test/fuzz/urFuzz.cpp:5]: Done

File: test/fuzz/utils.hpp (hunk 16 above)
Comment (PatKamin) [reply test/fuzz/utils.hpp:111]: This way we would eventually get stuck with multiple calls to urPlatformGet(), as it's the first item in enum representing 0. I prefer to have it this way and terminate the iteration on insufficient data.

File: .github/workflows/cmake.yml [hunk 23]
```diff
@@ -88,6 +88,60 @@ jobs:
...
+        -DUR_USE_UBSAN=ON
+        -DUR_DPCXX=${{github.workspace}}/dpcpp_compiler/bin/clang++
+
+    - name: Generate source from spec, check for uncommitted diff
+      run: cmake --build ${{github.workspace}}/build --target check-generated
+
+    - name: Build
+      run: cmake --build ${{github.workspace}}/build -j $(nproc)
+
+    - name: Fuzz long test
```
Comment (lukaszstolarczuk) [inline .github/workflows/cmake.yml:141]: fuzz-long? 😉

File: .github/workflows/nightly.yml [hunk 24]
```diff
@@ -0,0 +1,60 @@
+name: Nightly
+
+on:
+  schedule:
+    - cron: '0 23 * * *'
```
Comment (lukaszstolarczuk) [inline .github/workflows/nightly.yml:5]: For debugging purpose, you can enable it temporarily for PR trigger (to see if it compiles properly)

Comment (PatKamin) [reply .github/workflows/nightly.yml]: Sure

File: test/fuzz/utils.hpp (hunk 11 above)
Comment (PatKamin) [reply test/fuzz/utils.hpp]: Done

File: test/fuzz/utils.hpp (hunk 10 above)
Comment (PatKamin) [reply test/fuzz/utils.hpp]: Added TODO

File: .github/workflows/cmake.yml (hunk 23 above)
Comment (PatKamin) [reply .github/workflows/cmake.yml:141]: Done

File: test/fuzz/utils.hpp (hunk 13 above)
Comment (PatKamin) [reply test/fuzz/utils.hpp]: Added try-catch

**Review by lukaszstolarczuk — COMMENTED** [review]
👍

File: test/fuzz/utils.hpp (hunk 17 above)
Comment (igchor) [reply test/fuzz/utils.hpp:79]: Ok, got it. Btw, you can also wrap `ur_context_handle_t` in unique_ptr and then you wouldn't need to implement the dtor at all. But both ways work

**Review by igchor — APPROVED** [review]

File: test/fuzz/corpus/README.md (hunk 19 above)
Comment (pbalcer) [reply test/fuzz/corpus/README.md:4]: https://github.com/google/fuzzing/blob/master/tutorial/libFuzzerTutorial.md#seed-corpus

There's still some work to be done to make a good, minimized, seed corpus. But that's work for later. At that point, we should definitely expand this documentation.

**Review by pbalcer — APPROVED** [review]
